### PR TITLE
WEB-7869 - Added extra helper to filter featured items within an ordered list

### PR DESCRIPTION
**Pull request by owennicol**
#### What are the links to relevant tickets?
https://hxshortbreaks.atlassian.net/browse/WEB-7869

#### What does this PR do?
Added additional dust helper to allow an ordered list to be filtered. e.g. a list of hotels with 'featured' hotels removed or a list of just featured hotels.

#### What functional/unit tests does this PR have?
No additional
#### Is there anything a developer should be aware of when reviewing this?
This relates to work on Theatre Blueprint:
https://github.com/holidayextras/theatreBlueprint/pull/186

#### Screenshots (if appropriate)
---
#### Quality checklist (for PR author to complete BEFORE code review):
- [x] I've checked there is appropriate unit test coverage.
- [x] I've updated documentation with any changes to procedures.
- [x] I've checked this work against the requirements of the Jira.
- [x] This change passes all necessary tests (cross browser, automated or otherwise).
- [x] I am ready for this to be code reviewed, merged and tested.

#### Reviewer 1 @kevinhodges
- [x] I agree with the assumptions made in the quality checklist.
- [x] I’ve witnessed the work behaving as expected.
- [x] I’ve checked for appropriate test coverage.
- [x] I’ve checked for coding anti-patterns.
- [x] I've checked this work against the requirements of the Jira.
- [x] I don’t believe this work introduces unnecessary technical debt.
- [x] +1 from me!

#### Reviewer 2 @taijuten
- [x] I agree with the assumptions made in the quality checklist.
- [x] I’ve witnessed the work behaving as expected.
- [x] I’ve checked for appropriate test coverage.
- [x] I’ve checked for coding anti-patterns.
- [x] I've checked this work against the requirements of the Jira.
- [x] I don’t believe this work introduces unnecessary technical debt.
- [x] +1 from me!

### CODE REVIEW

File: jsconfig.json [hunk 1]
```diff
@@ -0,0 +1,16 @@
+{
+	// See https://go.microsoft.com/fwlink/?LinkId=759670
+	// for the documentation about the jsconfig.json format
```
Comment (notABluesSinger) [inline jsconfig.json]: You can't have comments in JSON

File: package.json [hunk 2]
```diff
@@ -1,7 +1,7 @@
 {
   "name": "the-wall-templater-dustjs",
-  "description": "The Wall compatible template handler",
-  "version": "3.1.1",
+  "description": "The Wall compatible template handler ",
```
Comment (notABluesSinger) [inline package.json]: Don't need this extra space

File: jsconfig.json (hunk 1 above)
Comment (notABluesSinger) [inline jsconfig.json]: Also, what is this file for, is it from your editor, if so, can you add the filename into the `.gitignore`

File: helpers/sortAndFilterHotels.js [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+   * @param {string} node Child node to sort object by
+   * @param {object} filterArray - array to base the filter on
+   * @param {boolean as string} exclude - if 'true', the filter function will exclude all items in the filterArray, based on the filterValue
+   * @param {boolean as string} include - if 'true', the filter function will include all items in the filterArray
+   * @example {@_sortAndFilterHotels parent=packageRatesReply.packageRates node="grossPrice" filterArray=_brandConfig.featuredHotel filterValue="x.links.hotelProducts.ids[0]" include="true"} {/_sortAndFilterHotels} output loop of parent object sorted by node and filtered by filterArray
+   */
+
+  dust.helpers._sortAndFilterHotels = function(chunk, context, bodies, params) {
+    var parent = params.parent;
+    var node = params.node;
```
Comment (notABluesSinger) [inline helpers/sortAndFilterHotels.js]: As both of these variables are only used once and the name isn't really adding more context, this isn't worth declaring. You may as well reference directly on L28

File: helpers/sortAndFilterHotels.js [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+    var filterArray = params.filterArray;
+    var include = params.include;
+    var exclude = params.exclude;
```
Comment (notABluesSinger) [inline helpers/sortAndFilterHotels.js]: Same for `include` and `exclude`... not really worth declaring

File: helpers/sortAndFilterHotels.js [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+
+    var mappedObject = sortBy(parent, node);
+    var newMappedObject = [];
+
+    //if filterArray exists, filter out from mappedObject
+    if (filterArray.length) {
+      if (exclude) {
+        newMappedObject = _.filter(mappedObject, function(x) {
```
Comment (notABluesSinger) [inline helpers/sortAndFilterHotels.js]: Rather than `x` can we use a meaningful variable name.

File: helpers/sortAndFilterHotels.js [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+module.exports = function(dust) {
+
+  /*
+   * @disclaimer - this is a temporary filter helper to separate a recommended hotel(s) from a list of hotels. This is temporary because it's written in dust, which will change soon.
+   *
+   * @description Augmented sorting method that orders a list of things based on a child node, and then filters the result based on an object and a value .
+   * @param {object} parent Object to sort
+   * @param {string} node Child node to sort object by
+   * @param {object} filterArray - array to base the filter on
+   * @param {boolean as string} exclude - if 'true', the filter function will exclude all items in the filterArray, based on the filterValue
```
Comment (kevinhodges) [inline helpers/sortAndFilterHotels.js]: would it make sense to have one variable for both `include` and `exclude` and simply flip the negative? I'm guessing you'd never use both at the same time, the only issue would be if you can pass neither when calling?

File: helpers/sortAndFilterHotels.js [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+
+  dust.helpers._sortAndFilterHotels = function(chunk, context, bodies, params) {
+    var filterArray = params.filterArray;
+
+    var mappedObject = sortBy(params.parent, params.node);
+    var newMappedObject = [];
+
+    //if filterArray exists, filter out from mappedObject
+    if (filterArray.length) {
+      if (params.exclude) {
```
Comment (kevinhodges) [inline helpers/sortAndFilterHotels.js]: There's a few lines of very similar code here, could we refactor to something like this?
```
if (params.include || params.exclude) {
	newMappedObject = _.filter(mappedObject, function(object) {
		//filter mapped object, exclude all recommended hotels from hotels list
		var hotelIndex = filterArray.indexOf(object.links.hotelProducts.ids[0]); 
		if (params.include) {
			return hotelIndex > -1;
		}
		//exclude logic...
		return hotelIndex < 0;
	});
}
```
Breaks the logic checking piece (`>` or <`) apart too

File: helpers/sortAndFilterHotels.js (hunk 6 above)
Comment (owennicol) [inline helpers/sortAndFilterHotels.js]: I've implemented this

File: helpers/sortAndFilterHotels.js [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+  dust.helpers._sortAndFilterHotels = function(chunk, context, bodies, params) {
+    var filterArray = params.filterArray;
+
+    var mappedObject = sortBy(params.parent, params.node);
+    var newMappedObject = [];
+
+    //if filterArray exists, filter out from mappedObject
+    if (filterArray.length && params.isIncludedInArray) {
+      newMappedObject = _.filter(mappedObject, function(object) {
+        //filter mapped object, exclude all recommended hotels from h(otels list
```
Comment (taijuten) [inline helpers/sortAndFilterHotels.js]: small typo

File: helpers/sortAndFilterHotels.js [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+        var hotelIndex = filterArray.indexOf(object.links.hotelProducts.ids[0]);
+        return (params.isIncludedInArray === 'true') ? hotelIndex > -1 : hotelIndex < 0;
+      });
+    }
+
+    if (newMappedObject.length) {
+      mappedObject = newMappedObject;
+    }
+    _.forEach(mappedObject, function(item) {
```
Comment (taijuten) [inline helpers/sortAndFilterHotels.js]: could this be a `return _.map` instead of a `_.forEach` then a `return`?

File: package.json [hunk 10]
```diff
@@ -1,7 +1,7 @@
 {
   "name": "the-wall-templater-dustjs",
   "description": "The Wall compatible template handler",
-  "version": "3.1.1",
+  "version": "3.1.7",
```
Comment (taijuten) [inline package.json]: strange number of version bumps, have there been a load of patches to staging?

Comment (owennicol) [inline package.json]: yes, there were a few very minor changes that I bumped the version for, in hindsight I wouldn't have done

**Pull request merged**